panel/links.py

```diff
@@ -1,21 +1,40 @@
 """
 Defines Links which allow declaring links between bokeh properties.
 """
+from __future__ import annotations
+
 import difflib
 import sys
+import warnings
 import weakref
 
-import param
+from typing import (
+    TYPE_CHECKING, Any, Dict, Iterable, List, Optional, Sequence, Tuple, Type, Union
+)
+
+from bokeh.models import CustomJS, LayoutDOM
+from bokeh.models import Model as BkModel
 
-from bokeh.models import CustomJS, Model as BkModel, LayoutDOM
+import param
 
 from .io.datamodel import create_linked_datamodel
 from .models import ReactiveHTML
 from .reactive import Reactive
 from .viewable import Viewable
 
+if TYPE_CHECKING:
+    from bokeh.model import Model
+
+    try:
+        from holoviews.core.dimension import Dimensioned
+        JSLinkTarget = Union[Reactive, BkModel, 'Dimensioned']
+    except Exception:
+        JSLinkTarget = Union[Reactive, BkModel] # type: ignore
+    SourceModelSpec = Tuple[Optional[str], str]
+    TargetModelSpec = Tuple[Optional[str], Optional[str]]
 
-def assert_source_syncable(source, properties):
+
+def assert_source_syncable(source: 'Reactive', properties: Iterable[str]) -> None:
```

Comment:
When using `from __future__ import annotations`, i.e. lazy evaluation of annotations, parentheses aren't needed anymore (in case of own or forward references, etc.).

Reply:
Yeah, gonna clean that up consistently at some point.